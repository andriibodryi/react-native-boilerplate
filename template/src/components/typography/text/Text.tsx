import * as React from 'react';

import { useStyleConfig } from '@src/styles/useStyleConfig';

import { BasicText } from '../basic-text';
import { TextTypes } from './text.types';
import { presets, textDecorators } from './text.presets';

const Text: React.FC<TextTypes> = ({
  preset = 'textMdRegular',
  style,
  appearance = 'defaultDecoration',
  color = 'primary',
  ...rest
}) => {
  const { colors } = useStyleConfig();

  const textColor = {
    primary: colors.grey_03,
    secondary: colors.grey_07,
  };

  return (
    <BasicText
      style={[
        presets[preset],
        textDecorators[appearance],
        {
          color: textColor[color],
        },
        style,
      ]}
      {...rest}
    />
  );
};

export default Text;
