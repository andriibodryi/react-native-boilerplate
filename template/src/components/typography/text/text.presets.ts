import { FONT_SIZE, FONT_WEIGHT, LINE_HEIGHT, MAIN_FONT } from '../fonts';

const BASIC_REGULAR = {
  fontWeight: FONT_WEIGHT.regular,
  fontFamily: MAIN_FONT.regular,
};

const BASIC_MEDIUM = {
  fontWeight: FONT_WEIGHT.medium,
  fontFamily: MAIN_FONT.medium,
};

export const presets = {
  textMdRegular: {
    ...BASIC_REGULAR,
    fontSize: FONT_SIZE.md,
    lineHeight: LINE_HEIGHT.md,
  },
  textMdLineXlRegular: {
    ...BASIC_REGULAR,
    fontSize: FONT_SIZE.md,
    lineHeight: LINE_HEIGHT.xl,
  },
  textMdMedium: {
    ...BASIC_MEDIUM,
    fontSize: FONT_SIZE.md,
    lineHeight: LINE_HEIGHT.md,
  },
  textMdLineXlMedium: {
    ...BASIC_MEDIUM,
    fontSize: FONT_SIZE.md,
    lineHeight: LINE_HEIGHT.xl,
  },
  textMdLine2XlMedium: {
    ...BASIC_MEDIUM,
    fontSize: FONT_SIZE.md,
    lineHeight: LINE_HEIGHT['2xl'],
  },
  textMdLine3XlMedium: {
    ...BASIC_MEDIUM,
    fontSize: FONT_SIZE.md,
    lineHeight: LINE_HEIGHT['3xl'],
  },
  textSmRegular: {
    ...BASIC_REGULAR,
    fontSize: FONT_SIZE.sm,
    lineHeight: LINE_HEIGHT.sm,
  },
  textSmMedium: {
    ...BASIC_MEDIUM,
    fontSize: FONT_SIZE.sm,
    lineHeight: LINE_HEIGHT.sm,
  },
  textXsRegular: {
    ...BASIC_REGULAR,
    fontSize: FONT_SIZE.xs,
    lineHeight: LINE_HEIGHT.xs,
  },
  textXsMedium: {
    ...BASIC_MEDIUM,
    fontSize: FONT_SIZE.xs,
    lineHeight: LINE_HEIGHT.xs,
  },
  textXlMedium: {
    ...BASIC_MEDIUM,
    fontSize: FONT_SIZE['2xl'],
    lineHeight: LINE_HEIGHT['2xl'],
  },
};

export const textDecorators = {
  underline: {
    textDecorationLine: 'underline',
  },
  defaultDecoration: {},
};
