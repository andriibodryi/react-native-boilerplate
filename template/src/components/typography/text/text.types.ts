import React from 'react';
import { Text as NativeText, TextStyle, StyleProp } from 'react-native';

import { presets, textDecorators } from './text.presets';

export type TextPreset = keyof typeof presets;

type TextAppearance = keyof typeof textDecorators;

export type TextColor = 'primary' | 'secondary';

export type TextTypes = React.ComponentProps<typeof NativeText> & {
  preset?: TextPreset;
  style?: StyleProp<TextStyle>;
  appearance?: TextAppearance;
  color?: TextColor;
};
