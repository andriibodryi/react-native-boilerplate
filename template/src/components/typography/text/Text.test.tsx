import React from 'react';
import { render } from '@testing-library/react-native';

import Text from './Text';
import { presets } from './text.presets';

const setup = (
  overrideOrAddProps?: Partial<React.ComponentProps<typeof Text>>,
) => render(<Text {...overrideOrAddProps} />);

describe('Text', () => {
  it('should render correctly when passing no props', () => {
    const { toJSON } = setup();

    expect(toJSON()).toMatchSnapshot();
  });

  describe('presets', () => {
    const mockText = 'Lorem ipsum dolor sit amet consectetur adipisicing elit.';

    it('should accept no preset - defaults to textMdRegular', () => {
      const { getByText, toJSON } = setup({
        children: mockText,
      });

      expect(getByText(mockText).props.children).toBe(mockText);
      expect(getByText(mockText).props.style.flat()).toContain(
        presets.textMdRegular,
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should accept buttonSecondary as preset', () => {
      const { getByText, toJSON } = setup({
        children: mockText,
        preset: 'textSmRegular',
      });

      expect(getByText(mockText).props.children).toBe(mockText);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should be resilient to invalid presets', () => {
      const { getByText, toJSON } = setup({
        children: mockText,
        preset: 'textMdMedium',
      });

      expect(getByText(mockText).props.children).toBe(mockText);
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
