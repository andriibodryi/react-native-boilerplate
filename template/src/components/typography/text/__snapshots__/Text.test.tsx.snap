// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text presets should accept buttonPrimary as preset 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "fontFamily": "Aspira-Medium",
          "fontSize": 12,
          "fontWeight": "500",
          "letterSpacing": 1.5,
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.
</Text>
`;

exports[`Text presets should accept buttonSecondary as preset 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "fontFamily": "Aspira-Regular",
          "fontSize": 14,
          "fontWeight": "400",
          "lineHeight": 18,
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.
</Text>
`;

exports[`Text presets should accept heading2 as preset 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "color": "#4D4D4D",
          "fontFamily": "Aspira-Medium",
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.
</Text>
`;

exports[`Text presets should accept iconText as preset 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "fontFamily": "Aspira-Demi",
          "fontSize": 18,
          "fontWeight": "500",
          "lineHeight": 20,
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.
</Text>
`;

exports[`Text presets should accept no preset - defaults to textMdRegular 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "fontFamily": "Aspira-Regular",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 20,
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.
</Text>
`;

exports[`Text presets should accept tabText as preset 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "fontFamily": "Aspira-Medium",
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.
</Text>
`;

exports[`Text presets should be resilient to invalid presets 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "fontFamily": "Aspira-Medium",
          "fontSize": 16,
          "fontWeight": "500",
          "lineHeight": 20,
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.
</Text>
`;

exports[`Text should render correctly when passing no props 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D4D4D",
        "fontFamily": "Aspira-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 20,
      },
      Array [
        Object {
          "fontFamily": "Aspira-Regular",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 20,
        },
        Object {},
        Object {
          "color": "#4D4D4D",
        },
        undefined,
      ],
    ]
  }
/>
`;
